### PR TITLE
Run tests under Valgrind to detect memory issues

### PR DESCRIPTION
**Pull request by javierhonduco**
Adding Valgrind could help us detect memory leaks and memory unsafety. Assuming Valgrind is installed, it can be run with:

```
$ cargo build --example smoketest
$ valgrind --version
valgrind-3.15.0
$ sudo valgrind --suppressions=build/valgrind-suppressions.supp target/debug/examples/smoketest     
```

This currently suppress the error below, to make it easier to detect new issues that are introduced. I am unsure if it's a false positive, will try to investigate later
```
[...]
smoketest: empty table
==22911== Syscall param bpf(attr->key) points to unaddressable byte(s)
==22911==    at 0x6E84DDD: syscall (in /usr/lib64/libc-2.29.so)
==22911==    by 0x64CA8D0: bpf_map_get_next_key (in /usr/lib64/libbcc.so.0.12.0)
==22911==    by 0x64C7E03: bpf_get_first_key (in /usr/lib64/libbcc.so.0.12.0)
==22911==    by 0x11A5D2: bcc::table::EntryIter::start (table.rs:160)
==22911==    by 0x11A770: <bcc::table::EntryIter as core::iter::traits::iterator::Iterator>::next (dfa.rs:685)
==22911==    by 0x115AC1: find_at (vec.rs:2076)
==22911==    by 0x115AC1: find_at (exec.rs:590)
==22911==    by 0x115AC1: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter (re_bytes.rs:593)
==22911==    by 0x115E5A: find_at (dfa.rs:0)
==22911==    by 0x115E5A: find_at (exec.rs:590)
==22911==    by 0x115E5A: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter (re_bytes.rs:593)
==22911==    by 0x114099: find_at (iterator.rs:1671)
==22911==    by 0x114099: find_at (exec.rs:590)
==22911==    by 0x114099: core::iter::traits::iterator::Iterator::collect (re_bytes.rs:593)
==22911==    by 0x111E43: find_at (dfa.rs:0)
==22911==    by 0x111E43: find_at (exec.rs:590)
==22911==    by 0x111E43: smoketest::main (re_bytes.rs:593)
==22911==    by 0x114CCA: find_at (dfa.rs:785)
==22911==    by 0x114CCA: find_at (exec.rs:590)
==22911==    by 0x114CCA: std::rt::lang_start::{{closure}} (re_bytes.rs:593)
==22911==    by 0x13A077: {{closure}} (rt.rs:52)
==22911==    by 0x13A077: do_call<closure-0,i32> (panicking.rs:297)
==22911==    by 0x13A077: try<i32,closure-0> (panicking.rs:274)
==22911==    by 0x13A077: catch_unwind<closure-0,i32> (panic.rs:394)
==22911==    by 0x13A077: std::rt::lang_start_internal (rt.rs:51)
==22911==    by 0x114CA6: find_at (dfa.rs:875)
==22911==    by 0x114CA6: find_at (exec.rs:590)
==22911==    by 0x114CA6: std::rt::lang_start (re_bytes.rs:593)
==22911==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==22911== 
[...]
```

Should we add this in each action job or have one specific job with a LLVM and BCC version (i.e. the newer one that is supported) that runs Valgrind on the smoke tests and potentially some of the tools?

Thanks!

### CODE REVIEW

**Comment by brayniac** [comment]
I would propose running in each job (putting in the main `build/ci.sh`). The main length of the CI time is around building BCC from source, so it's likely most efficient to run this within each job. I also like that it would give more comprehensive coverage across the test matrix - while we *shouldn't* have leaks that are specific to one bcc version or llvm version, it's not entirely impossible that we could have one specific to a set of bcc versions. 

How much time would we add to the CI jobs by running for each of the tools too? In this repo, those have been selected to exercise and demonstrate particular functionality, so they're good targets for running valgrind too.

**Comment by javierhonduco** [comment]
> while we shouldn't have leaks that are specific to one bcc version or llvm version, it's not entirely impossible that we could have one specific to a set of bcc versions.

Very good point! 

> How much time would we add to the CI jobs by running for each of the tools too? In this repo, those have been selected to exercise and demonstrate particular functionality, so they're good targets for running valgrind too.

Yep! That's something that we should look into. Running the smoke tests / tools under Valgrind takes significantly longer than without, as expected, but it's still a reasonable amount of time (<1 minute in my laptop).

Sounds great! Let me ~~push force 20 times until I get the right GitHub Actions commands~~ iterate on this PR and see how timings will be affected

File: build/ci.sh [hunk 1]
```diff
@@ -29,6 +29,16 @@ sudo apt-get --yes install clang-"${LLVM_PACKAGE}" \
     libclang-"${LLVM_PACKAGE}"-dev libelf-dev libfl-dev \
     llvm-"${LLVM_PACKAGE}"-dev libz-dev llvm-"${LLVM_PACKAGE}"
 
+sudo apt-get --yes install libc6-dbg
+
+pushd /tmp
+curl -L -O https://sourceware.org/pub/valgrind/valgrind-3.16.1.tar.bz2
+tar xvjf valgrind-3.16.1.tar.bz2
```
Comment (brayniac) [inline build/ci.sh:36]: Maybe you just had this for testing, but I don't think we need verbose output from tar here.

**Comment by brayniac** [comment]
Probably need to add the `.swp` to gitignore and remove it from this PR.

**Comment by javierhonduco** [comment]
Thanks for taking a look! I'll mark it for review as soon as I see that CI passes cleanly. Right now there are 3 Valgrind suppressions, and two of them come for LLVM / Clang, which I don't think are this project's fault, but I hope I will find the time to take a look at the bpf syscall with the `NULL` argument, which repros in my dev box!

**Comment by javierhonduco** [comment]
Hope I haven't missed any suppression 😄

**Comment by brayniac** [comment]
Sounds good. Just request review / @ me here when you're happy with the PR.

Really appreciate you taking the time to add this. It's going to be a big help in making sure we do the right things in this crate.

**Comment by javierhonduco** [comment]
Hope this will catch bugs earlier, faster! The codebase and testing setup made it way easier, so thanks! 😄

**Review by brayniac — APPROVED** [review]

**Pull request merged**